### PR TITLE
Feature/v10/list item

### PR DESCRIPTION
**Pull request by rloos289**
This is to integrate bootstrap group-items and badges into PL.

### CODE REVIEW

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        <div class="mb-3">
+          {% include '@molecules/card/_card.twig' with list_group_helper_toggleable %}
+        </div>
+        <div class="row mb-3">
+          <div class="col-sm">
+            <h2>Toggleable</h2>
+            {% include '@atoms/list-group/_list-group.twig' with list_group_toggleable %}
+          </div>
+          <div class="col-sm">
+            <div class="tab-content" id="nav-tabContent">
```
Comment (rloos289) [inline source/_patterns/01-atoms/list-group/demo/list-groups.twig]: I ended up manually coding the tab list because our list component wasn't quite right for it. The list components all start with a 'list-group-item' class and these do not.

File: source/_patterns/00-base/_transitions.scss [hunk 2]
```diff
@@ -6,5 +6,6 @@
 //$transition-fade
 //$btn-transition
 @import '~bootstrap/scss/forms';
```
Comment (illepic) [inline source/_patterns/00-base/_transitions.scss]: @rloos289, this is a copy/paste error and will result in duplicate form output. Kill it.

File: source/_patterns/01-atoms/badge/demo/badges.twig [hunk 3]
```diff
@@ -0,0 +1,53 @@
+<div class="container">
+  <div class="row">
+    {% embed '@atoms/grid/_grid-1-up.twig' %}
```
Comment (illepic) [inline source/_patterns/01-atoms/badge/demo/badges.twig]: No need for `.row` here, it's included in `_grid-1-up.twig`

File: source/_patterns/01-atoms/badge/demo/badges.twig [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+          <p>Note: Badges scale to match the size of the immediate parent element </p>
+          {% for size in sizes %}
+            <{{ size.badge_size }}>{% include '@atoms/badge/_badge.twig' with size %}</{{ size.badge_size }}>
+          {% endfor %}
+        </div>
+
+        <div class="mb-3">
+          <h1>Badge Button</h1>
+          {% embed '@atoms/button/_button.twig' with {
+          button_color: 'primary',
```
Comment (illepic) [inline source/_patterns/01-atoms/badge/demo/badges.twig]: formatting here (indent vars)

File: source/_patterns/01-atoms/badge/demo/badges.yml [hunk 5]
```diff
@@ -0,0 +1,109 @@
+badge_helper:
+  card_title: Badges
+  card_text: "
+             <p><code>badge_element</code> String: span|a choose the element type for your badge. If you use an link element make sure to enter a value for badge_link</p>
```
Comment (illepic) [inline source/_patterns/01-atoms/badge/demo/badges.yml]: Maybe just include this in the HTML of the demo? Does it really need to be in the yaml?

File: source/_patterns/01-atoms/badge/demo/badges.yml [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    badge_text: Badge
+    badge_size: h5
+  - badge_color: primary
+    badge_text: Badge
+    badge_size: h6
+
+button_badge:
+  button_color: primary
+  button_text: Notifications
+colors:
```
Comment (illepic) [inline source/_patterns/01-atoms/badge/demo/badges.yml]: What if we just had a universal set of values on `_data.json`? ie:

```json
{
  "colors": [ "primary", "secondary", "etc"]
}
```

Do the same thing for sizes, etc from Bootstrap. Looping through the same "things" in every demo doesn't speak to the "systems" at play across the board.

File: source/_patterns/01-atoms/badge/demo/badges.yml [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  - badge_color: info
+    badge_text: Info
+    badge_pill: TRUE
+  - badge_color: light
+    badge_text: Light
+    badge_pill: TRUE
+  - badge_color: dark
+    badge_text: Dark
+    badge_pill: TRUE
+links:
```
Comment (illepic) [inline source/_patterns/01-atoms/badge/demo/badges.yml]: And technically this can just be a loop of global colors, with hard-coded values in the demo.

File: source/_patterns/01-atoms/button/_button.twig [hunk 8]
```diff
@@ -36,4 +36,10 @@
   button_disabled ? 'disabled',
 ] | sort | join(' ') %}
 
-<{{ button_element }} type="{{ button_type }}" class="{{ classes }}" {{ attributes }}> {{ button_text }} </{{ button_element }}>
+<{{ button_element }} type="{{ button_type }}" class="{{ classes }}" {{ attributes }}>
+{{ button_text }}
+{% block right %}
+
+{% endblock right %}
```
Comment (illepic) [inline source/_patterns/01-atoms/button/_button.twig]: Go ahead and move this up linline to opening block statement.

File: source/_patterns/01-atoms/list-group/demo/list-groups.md [hunk 9]
```diff
@@ -0,0 +1,11 @@
+## Variables
+
+Name  | Type | Options | Description
```
Comment (madhaze) [inline source/_patterns/01-atoms/list-group/demo/list-groups.md]: Looks like this table items needs starting `|`s

File: source/_patterns/01-atoms/list-group/index.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+// import $ from 'jquery';
+
+// Module dependencies
+import base from 'base';
+
+// Module styles
+import './_list-group.scss';
+
+export default {
+  name: 'list-item',
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/index.js]: This is named `list-item` though it is in `list-group` folder.

File: source/_patterns/01-atoms/badge/demo/badges.twig [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+        {% embed '@atoms/button/_button-outline.twig' with {
+          button_color: 'primary',
+          button_text: 'Notifications',
+        }%}
+          {% block right %}
+            {% include '@atoms/badge/_badge.twig' with {
+            badge_text: '4',
+            badge_color: 'primary',
+
```
Comment (madhaze) [inline source/_patterns/01-atoms/badge/demo/badges.twig]: extra space here

File: source/_patterns/01-atoms/badge/demo/badges.md [hunk 12]
```diff
@@ -0,0 +1,9 @@
+## Variables
+
+Name  | Type | Options | Description
+--- | --- | --- | ---
+badge_element | string | span, a | choose the element type for your badge. If you use a link tag make sure to enter a value for badge_link
```
Comment (illepic) [inline source/_patterns/01-atoms/badge/demo/badges.md]: Is this missing the left-side `|` required for markdown tables?

File: source/_patterns/01-atoms/button/demo/buttons.twig [hunk 13]
```diff
@@ -1,23 +1,33 @@
...
+        } %}
       {% endfor %}
     </div>
 
-    <h1>Button outline</h1>
+    <h1>Button elements</h1>
     <div class="mb-3">
-      {% for outline in outlines %}
-        {% include '@atoms/button/_button.twig' with outline %}
+      {% for element in elements %}
```
Comment (illepic) [inline source/_patterns/01-atoms/button/demo/buttons.twig]: This really generic global(?) variable? Where does this come from? And can we change it to something with less of a potential for naming collision?

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+          card_title: 'Basic Examples',
+          card_text: list_group_helper_basic,
+        } %}
+      </div>
+      <div class="row mb-3">
+        <div class="col-sm-6">
+          {% include '@atoms/list-group/_list-group.twig' with list_group_text %}
+        </div>
+        <div class="col-sm-6">
+          {# This block of code is going to appear multiple times. It exists because the _list-group.twig template ultimately needs an object to iterate through to generate the individual items. These code blocks exist to pull data from multiple disperate places (the all_colors array, list-groups.yml, etc.) and combine them into a single object that the template can then process. #}
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/demo/list-groups.twig]: Let's set comment width to 100 chars max.

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+          {% set list_color = [] %}
+          {% for item in list_group_text.items %}
+            {% set temp_item = [] | merge(list_group_text.items[loop.index0]) | merge(all_colors[loop.index0]) %}
+            {% set list_color = list_color|merge([temp_item]) %}
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/demo/list-groups.twig]: This seems unnecessary, seems like it can be integrated into the line above.

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+          card_text: list_group_helper_active,
+        } %}
+      </div>
+      <div class="row mb-3">
+        <div class="col-sm">
+          <h2>Active Items</h2>
+          {% set list_active = [] %}
+          {% for item in list_group_text.items %}
+            {% set temp_item = [] | merge(list_group_text.items[loop.index0]) | merge(list_group_active.items[loop.index0]) %}
+            {% set list_active = list_active|merge([temp_item]) %}
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/demo/list-groups.twig]: See note above.

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+          card_title: 'Adding Elements',
+          card_text: list_group_helper_adding,
+        } %}
+      </div>
+      <div class="row mb-3">
+        <div class="col-sm">
+          <h2>Badges</h2>
+          {% set list_badge = [] %}
+          {% for item in list_group_text.items %}
+            {% set temp_item = [] | merge(list_group_text.items[loop.index0]) | merge(all_colors[loop.index0]) | merge(list_group_badges.items[0]) | merge(list_group_badges.items[1])%}
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/demo/list-groups.twig]: Let's break these into multiple lines.

File: source/_patterns/01-atoms/list-group/index.js [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+
+// import $ from 'jquery';
+
+// Module dependencies
+import base from 'base';
+
+// Module styles
+import './_list-group.scss';
+
+export default {
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/index.js]: We don't use this module format any longer. Pull from `epic/v10` for shape of this file.

File: source/_patterns/02-molecules/card/_card-helper.twig [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+
+  This is a simple example of extending an existing template to make a new template
+#}
+
+{% extends '@molecules/card/_card.twig' %}
+{# In this context we're hard-coding the only thing we need to override for a helper card. #}
+{% set text_color = 'light' %}
+{% set card_background = 'secondary' %}
+{% set button_element = 'a' %}
+{% set button_text = 'More Info' %}
```
Comment (illepic) [inline source/_patterns/02-molecules/card/_card-helper.twig]: Pipe through `|t`

File: source/_patterns/02-molecules/card/_card.twig [hunk 20]
```diff
@@ -43,66 +39,62 @@
...
-          button_text: button_text,
-          button_color: button_color,
-          button_link: button_link,
-        } only %}
-      </div>
-    {% endif %}
+
+  {% if card_flush %}
+    {# Note that 'p2card--flush--body' is not a bootstrap class. It only being used to designate the flush body#}
+    {% set body_class = 'p2card--flush--body' %}
```
Comment (illepic) [inline source/_patterns/02-molecules/card/_card.twig]: This isn't a BEM name. Does this class just not apply padding?

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig (hunk 15 above)
Comment (rloos289) [reply source/_patterns/01-atoms/list-group/demo/list-groups.twig]: Could you clarify how to do that? . The only way I can think to do that would be something like 
`{% set list_color[] = [] | merge(list_group_text.items[loop.index0]) | merge(all_colors[loop.index0]) %}` so that its adding to the array and not overriding it in each iteration of the loop. But this syntax doesn't work.

Comment (illepic) [reply source/_patterns/01-atoms/list-group/demo/list-groups.twig]: @rloos289, let's do this: Can you distill down the data structures via https://twigfiddle.com/ so I can go hands-on with it?

File: source/_patterns/02-molecules/card/_card.twig (hunk 20 above)
Comment (anniegreens) [reply source/_patterns/02-molecules/card/_card.twig]: Would it be more applicable to make this a utility class that could be used either: a) only within PL; or b) globally by the system (possibly tied to something BS provides). I'm still not sure there's a need for a BEM style class at all here if it is only for demo purposes (if I'm understanding why it is being used).

File: source/_patterns/01-atoms/list-group/_list-group.twig [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+
+<{{ parent_element }} class="list-group {{ list_flush }}" id="{{ list_id }}">
+  {% for item in items %}
+
+  {# Note: ?: won't work in PHP5 #}
+  {% set element = item.item_element ?: 'li' %}
+
+  {% set classes = [
+  'list-group-item',
+  'list-group-item-' ~ item.item_color,
```
Comment (rloos289) [inline source/_patterns/01-atoms/list-group/_list-group.twig]: space here

File: source/_patterns/01-atoms/list-group/_list-group.twig [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+#}
+
+{% set parent_element = parent_element ?: 'ul' %}
+
+{% if list_flush %}
+  {% set list_flush = 'list-group-flush' %}
+{% endif %}
+
+<{{ parent_element }} class="list-group {{ list_flush }}" id="{{ list_id }}">
+  {% for item in items %}
```
Comment (rloos289) [inline source/_patterns/01-atoms/list-group/_list-group.twig]: redo id so that if theres no id, the html element doesn't appear at all

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig (hunk 15 above)
Comment (rloos289) [reply source/_patterns/01-atoms/list-group/demo/list-groups.twig]: Maybe refactor this whole thing using macros instead?

**Comment by illepic** [comment]
Getting this error on this branch:

2017-11-27 at 11 18

**Comment by illepic** [comment]
Also, let's ditch the dark wells around the demos:

2017-11-27 at 11 25

It's a little too much going on around the stuff we're trying to show.

File: source/_patterns/01-atoms/badge/demo/badges.md [hunk 23]
```diff
@@ -0,0 +1,9 @@
+## Variables
+
+| Name  | Type | Options | Description
+--- | --- | --- | ---
+badge_element | string | span, a | choose the element type for your badge. If you use a link tag make sure to enter a value for badge_link
```
Comment (illepic) [inline source/_patterns/01-atoms/badge/demo/badges.md]: Let's get the left-bars in for the markdown table syntax (was mentioned before)

File: source/_patterns/01-atoms/list-group/_list-group.scss [hunk 24]
```diff
@@ -4,5 +4,6 @@
 // copy from ~bootstrap/scss/variables to here
 
 @import "~bootstrap/scss/list-group";
+//@import "~bootstrap/scss/transitions";
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/_list-group.scss]: Nuke this.

File: source/_patterns/01-atoms/list-group/_list-group.twig (hunk 21 above)
Comment (illepic) [reply source/_patterns/01-atoms/list-group/_list-group.twig]: Have to revisit this.

File: source/_patterns/01-atoms/list-group/demo/list-groups.md [hunk 25]
```diff
@@ -0,0 +1,11 @@
+## Variables
+
+| Name  | Type | Options | Description
+--- | --- | --- | ---
+parent_element | string | ul,ol,div | the html element that will contain the list
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/demo/list-groups.md]: Left bars.

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig [hunk 26]
```diff
@@ -0,0 +1,209 @@
+{# This function loops through the scssColors global variable to make an array of all the available colors. #}
+{% set colors = [] %}
+{% for name, rgb in scssColors %}
+  {% set item_classes = {
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/demo/list-groups.twig]: Why is this here? Delete.

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+      <div class="mb-3">
+        <h1 class="text-center">List Group</h1>
+        {% include '@molecules/card/_card-helper.twig' with {
+          card_text: list_group_helper,
+        } %}
+      </div>
+
+      {# Basic examples #}
+      <div class="mb-3">
+        {% include '@molecules/card/_card-helper.twig' with {
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/demo/list-groups.twig]: Want to remove `card-helper` styling wrapper styles right before each thing we demo. Too much dark filling up the page. We can leave the first, overall description (ie line 16).

File: source/_patterns/01-atoms/list-group/demo/list-groups.twig [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+        <div class="col-sm-6">
+          {# This block of code is going to appear multiple times. It exists because the _list-group.twig
+          template ultimately needs an object to iterate through to generate the individual items. These code
+          blocks exist to pull data from multiple disperate places (the all_colors array, list-groups.yml,
+          etc.) and combine them into a single object that the template can then process. #}
+          {% set color_items = [] %}
+          {# Iterate over colors, since this is actually what we're demo'ing #}
+          {% for color in colors %}
+            {# Inline object merged into random text object #}
+            {% set color_item = { item_color: color } | merge(random(list_group_text.items)) %}
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/demo/list-groups.twig]: If everything else is going to use `temp_item` instead of their specific case (ie `active_item`, `disabled_item`, `action_item`) then ensure all are consistent. So change this `color_item` (or the others) to be a consistent naming pattern.

File: source/_patterns/01-atoms/list-group/index.js [hunk 29]
```diff
@@ -0,0 +1,19 @@
+/**
+ * List-group
+ */
+import 'bootstrap/js/src/tab';
```
Comment (illepic) [inline source/_patterns/01-atoms/list-group/index.js]: Noting here that we will probably build a standalone Tabs component and also require this there. Yay webpack!

File: source/_patterns/02-molecules/card/demo/cards.twig [hunk 30]
```diff
@@ -1,85 +1,133 @@
...
+    {% block column_1 %}
+      <h1 class="text-center">Cards</h1>
+      <div class="mb-3">
+        {% include '@molecules/card/_card-helper.twig' with {
+          card_text: helper_card_main
+        } %}
+      </div>
+      <div class="mb-3">
+        {% include '@molecules/card/_card-helper.twig' with {
+          card_title: "Text",
```
Comment (illepic) [inline source/_patterns/02-molecules/card/demo/cards.twig]: Single ticks, until we need string interpolation, then double ticks.
